### PR TITLE
Propagator using v5 solver

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Replacing the v4 solver for v5's one as the default solver in dev.major is still a ~40 files PR, so I am making the transition in block, starting by propagator and scattering.

**propagator.py:**
The function call the new solver and remove some options:
- batch: useless since __solve can evolve a propagator.
- parallel: Could still be useful, until we restore openmp support for data-layer, but should be managed by the data layer in v5.
- progress_bar: Now in options.
This result in a much simpler function.

Add a new class `Propagator`, which replace the `Evolver` in scattering.
Like solver classes, it split the setup and utilization:
```
U = Propagator(H)
psi_t = U(t) @ psi_0
psi_t2 = U(t2, t) @ psi_t
```
It save previous calls for efficiency, somewhat support `args`. It's not a `QobjEvo`, to support propagators with 2 times, but can be made one.

**scattering.py:**
- `Evolver` replaced by the previous `Propagator`.
- matrix-matrix product are replaced by matrix-vector product in photon scattering computation.
- Do no create a `Qobj` to extract one value from an array.
This result in a speed up of about 5 times faster for even small (N=10) operators.

**Changelog**
Rework propagator to use v5 solver and add a Propagator class.
Optimize scattering functions

### CODE REVIEW

**Comment by hodgestar** [comment]
@Ericgig This branch has some conflicts now. Could you merge dev.major into it?

File: qutip/solver/scattering.py [hunk 1]
```diff
@@ -109,53 +47,74 @@ def set_partition(collection, num_sets):
...
-
-    Returns
-    -------
-    omega : :class: qutip.Qobj
-        The temporal scattering operator with dimensionality equal to the
-        system state.
+    tlist : array_like
+        Times at which emissions can happen.
+    indices : list-like
+        List of (list of emission times indices) for each waveguide.
```
Comment (hodgestar) [inline qutip/solver/scattering.py:66]: I think the list of emission times for each waveguide was a better interface than these lists of indices into tlist. It's pretty easy to construct the tlist and the list of indices insides this function if we need them, but very annoying for the user to have to construct the tlist and the list of indices themselves every time they would like to calculate the scattering amplitudes.

File: qutip/solver/propagator.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            self.solver = SeSolver(Hevo, options=options)
+        self.cte = self.solver.rhs.isconstant
+        self.unitary = (not self.solver.rhs.issuper
+                        and isinstance(H, Qobj)
+                        and H.isherm)
+        self.args = args
+        self.memoize = max(3, memoize)
+        self.tol = tol
+
+    def __call__(self, t, t_start=0, **args):
```
[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/solver/propagator.py:176]: I first merged this but went back to `0`.  
Propagator do the evolution from the state at one time to the state at another, there are always 2 times involved.
While for constant system, only the difference is needed, so `t_start` is unneeded, for time dependent system `t_start` should be a valid time.

[the remark this answers is not in the document]
Comment (hodgestar) [reply qutip/solver/propagator.py:176]: `t_start=0` does make sense. :+1:

**Review by hodgestar — APPROVED** [review]
Looks good!

The documentation tests should likely pass first. I guess this is the same merge issue encountered in the qutip-qip removal PR.

**Pull request merged**